src/app/signup-form/signup-form.component.ts

```diff
@@ -0,0 +1,68 @@
+import { Component, OnInit } from '@angular/core';
+import { FormControl, FormGroup, FormGroupDirective, NgForm, Validators } from '@angular/forms';
+import { BirthdayValidator } from '../shared/directives/birthday-validator.directive';
+import { PasswordMatchValidator } from '../shared/directives/password-match-validator.directive';
+import { PasswordValidator } from '../shared/directives/password-validator.directive';
+
+@Component({
+  selector: 'app-signup-form',
+  templateUrl: './signup-form.component.html',
+  styleUrls: ['./signup-form.component.css']
+})
+export class SignupFormComponent implements OnInit {
+
+  signupForm!: FormGroup;
+  password!: FormGroup;
+  hideNew: boolean = true;
+  hideConfirm: boolean = true;
+  submitted: boolean = false;
+  emailError: boolean = false;
+
+  constructor() {
+    this.signupForm = new FormGroup({
+      'firstName': new FormControl('', [Validators.required]),
+      'lastName': new FormControl('', [Validators.required]),
+      'birthday': new FormControl('', [Validators.required, BirthdayValidator]),
+      'email': new FormControl('', [Validators.required, Validators.email]),
+      'password': new FormGroup({
+        'newPassword': new FormControl('', [Validators.required, PasswordValidator]),
+        'confirmNewPassword': new FormControl('', [Validators.required])
+      }, { validators: PasswordMatchValidator })
+    });
+  }
+
+  ngOnInit(): void {
+  }
+  get passwordGroup() {
+    return this.signupForm.get('password') as FormGroup;
```

Comment:
This is a good way to do this if you're looking to keep the HTML clean but you _can_ reference individual controls from HTML as well with `myForm.controls['controlName']` or `myForm.controls.controlName`